io-reactor/src/reactor.rs

```diff
@@ -369,33 +383,78 @@ impl<H: Handler, P: Poll> Runtime<H, P> {
+                                self.service.handle_listener_event(*id, event, time);
+                            }
+                        }
+                    }
+                    Err(IoFail::Connectivity(flags)) => {
+                        #[cfg(feature = "log")]
+                        log::trace!(target: "reactor", "Listener {id} hanged up (OS flags {flags:#b}");
+
+                        let listener = self.listeners.remove(id).expect("resource disappeared");
+                        self.service
+                            .handle_error(Error::ListenerDisconnect(*id, listener, flags));
+                    }
+                    Err(IoFail::Os(flags)) => {
+                        #[cfg(feature = "log")]
+                        log::trace!(target: "reactor", "Listener {id} errored (OS flags {flags:#b}");
+
+                        self.service
+                            .handle_error(Error::ListenerPollError(*id, flags));
                     }
                 }
             } else if let Some(id) = self.transport_map.get(&fd) {
-                #[cfg(feature = "log")]
-                log::trace!(target: "reactor", "Got `{io}` event from the transport {id} (fd={fd})");
+                match res {
+                    Ok(io) => {
+                        #[cfg(feature = "log")]
+                        log::trace!(target: "reactor", "Got `{io}` event from transport {id} (fd={fd})");
+
+                        let transport = self.transports.get_mut(id).expect("resource disappeared");
+                        for io in io {
+                            if let Some(event) = transport.handle_io(io) {
+                                self.service.handle_transport_event(*id, event, time);
+                            }
+                        }
+                    }
+                    Err(IoFail::Connectivity(flags)) => {
+                        #[cfg(feature = "log")]
+                        log::trace!(target: "reactor", "Transport {id} hanged up (OS flags {flags:#b}");
 
-                let res = self.transports.get_mut(id).expect("resource disappeared");
-                for io in io {
-                    if let Some(event) = res.handle_io(io) {
-                        self.service.handle_transport_event(*id, event, time);
+                        let transport = self.transports.remove(id).expect("resource disappeared");
+                        self.service
+                            .handle_error(Error::TransportDisconnect(*id, transport, flags));
+                    }
+                    Err(IoFail::Os(flags)) => {
+                        #[cfg(feature = "log")]
+                        log::trace!(target: "reactor", "Transport {id} errored (OS flags {flags:#b}");
+
+                        self.service
+                            .handle_error(Error::TransportPollError(*id, flags));
                     }
                 }
+            } else {
+                unreachable!()
```

Comment:
I think a panic or log is more appropriate here; we're not trying to prove anything to the compiler.

Reply:
Done